feat(cli-2.0): initial draft (#28)
appwrite/rfc: 018-cli-2.0/018-cli-2.0.md

> # Title <!-- What do you want to call your `awesome_feature`? -->

Reviewer: Missing title

* `appwrite login` makes a request to the existing `accounts.createSession` endpoint, captures the returned cookie and uses this cookie in all the subsequent requests. 

    Whenever we wish to access console specific features, we can authenticate those requests using the cookie and continue to use the API Key based authentication when using the CI.

    ```sh
    $ appwrite login
    Email:
    Password:
    ```

    In CI environments, the authentication will continue to take place using the API key on a per project basis until we develop a new token / key based mechanism for console authentication. 

    ```sh
    appwrite client --setKey="" --setEndpoint="" ...
    ```

* Second method relies on a browser based auth.

  ```sh
  $ appwrite login

  # Browser based authentication continues. After a successful login, the returned token and user ID are stored in a hidden preferences file.
  ```

  1. The command first finds an available local port. 
  2. Construct a redirect URL using the available port `https://localhost:1234`
  3. Spin up a local http server listening to on the available port ( 1234 in this case )

>   4. Opens the browser with a request to either the local appwrite server or appwrite cloud `/login?callback=https://localhost:1234` with the redirect URL as one of the query parameters.

Reviewer: ```suggestion
  4. Opens the browser with a request to either the local appwrite server or appwrite cloud `/authorize/cli?callback=https://localhost:1234` with the redirect URL as one of the query parameters.
```

This will also allow us to use `/authorize/vscode?callback=https://localhost:1234` in the future. We should also list some of the changes that will be required in the console itself. For example:
1. Custom redirect after login - if the user is not logged in the console, he should be redirect to the signin page, and after the signin he should be brought back to `/authorize/cli?callback=https://localhost:1234`.

2. We should also add protection against [open-redirect attack](https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html).

Author: done :+1:

  ![CLI Auth](cli_auth.png)

  This approach may seem unnecessarily complex now since we only allow email password based logins in the console. But as we move towards supporting OAuth logins in the console, this is the approach that needs to be followed. [Firebase CLI](https://github.com/firebase/firebase-tools/blob/0b0459bfde3fabf41223bb6d5d39a5cf325f1010/src/auth.ts#L476) for reference. 

## Easier Deployment of cloud functions
--- 

### 🟢 Constrain deployments from current directory.

We start by removing the `--code` parameter in the `createTag` command and constrain function deployments only from the current directory. 

The refactored command will look like this 

```sh
appwrite functions createTag --functionId=[ID] --command=[COMMAND]
```

This corresponding appwrite issue will also need to be tackled as we implement this.
https://github.com/appwrite/appwrite/issues/1316

> ### 🟢 The `.appwrite.json` File

Reviewer: Is there any benefit from a hidden file? Should we just use `appwrite.json`?

Author: I made it hidden under the assumption that we will need to store secrets in there. But since we planned to store the secrets in another file, this can be changed to  `appwrite.json`

The next improvement is going to be along the lines of a `.appwrite.json` file. Running `appwrite init` in the current directory will create a `.appwrite.json` file and initialize the current directory with attributes necessary to communicate with a particular project.

Consider this example
```sh
mkdir temp && cd temp

# Initialize the current directory with an Appwrite project 
appwrite init

temp 
└── .appwrite.json
```

A `.appwrite.json` file will be associated with one project only whose ID will be visible in the `project` attribute. 

```json
{
    "project" : "",
    "functions": [
        {
          "name" : "My awesome function 💪",
          "path" : "./FunctionOne",
          "command" : "python main.py",
          "runtime": "python3.9",
          // This section is useful when we allow cloud functions to be tested locally

>           "environment_variables" :{

Reviewer: ```suggestion
          "vars" :{
```

This is the name used here: https://appwrite.io/docs/server/functions#functionsCreate

            "KEY 1" : "VALUE 1",
          }
        },
    ],

>     "hosting": {}

Reviewer: ```suggestion
    "hosting": []
```

}
```

There are two scenarios here 
* A developer would like to link the directory to an existing project

```sh
appwrite init

Choose a project to link this directory to 
⭕️ Project 1
✅️ Project 2
⭕️ Project 3
```

* A developer would like to create a new project while initializing this directory

```sh
appwrite init --new

Give your project a name:
Project X

Creating Project X...

Project X Created 👍
```

> We need to first add support to utopia-php/CLI to allow for single select and multi select commands to enable this feature.

### 🟢 The `appwrite deploy` command

The `appwrite deploy` command is a convenient wrapper around three main commands
* Create Function
* Create Tag
* Update Tag

```sh
# Usage 1
appwrite deploy functions
Which functions would you like to deploy? 
✅️ FunctionOne
✅️ FunctionTwo
⭕️ FunctionThree

# Usage 2
appwrite deploy functions --all
# Deploys all the functions specified in .appwrite.json

# Future support for static hosting

> appwrite deploy site

Reviewer: We should choose a consistent name in the future (site/hosting).

Author: Yes sure :+1:

Here are some example usages 
```sh
# Usage 1
appwrite generate function --name="FunctionOne" --runtime=python-3.9
# Creates a folder called FunctionOne with a template for a sample python function.
temp
├── .appwrite.json
└── FunctionOne
    ├── .appwrite
    ├── main.py
    └── requirements.txt

# Usage 2
appwrite generate function --name="FunctionTwo" --runtime=node-16.0
# Creates a folder called FunctionTwo with a template for a sample node function.
temp 
├── .appwrite.json
├── FunctionOne
│   ├── .appwrite
│   ├── main.py
│   └── requirements.txt
└── FunctionTwo
    ├── index.js
    ├── node_modules
    └── package.json
```

Under the hood, this command does the following

1. Clones a publicly hosted git repository and copy's a particular folder into your working directory. This will allow us to fix / modify templates without creating a new CLI release 👍 

2. Makes an entry in `.appwrite.json`

```json
{
    "project" : "",
    "functions": [
        {
          "name" : "My awesome function 💪",
          "path" : "./FunctionOne",
          "command" : "python main.py",
          "runtime": "python3.9"
        },
    ],

>     "hosting": {}

Reviewer: ```suggestion
    "hosting": []
```

We start by removing the `--code` parameter in the `createTag` command and constrain function deployments only from the current directory. 

The refactored command will look like this 

```sh
appwrite functions createTag --functionId=[ID] --command=[COMMAND]
```

This corresponding appwrite issue will also need to be tackled as we implement this.
https://github.com/appwrite/appwrite/issues/1316

### 🟢 The `appwrite.json` File

The next improvement is going to be along the lines of a `appwrite.json` file. Running `appwrite init` in the current directory will create a `appwrite.json` file and initialize the current directory with attributes necessary to communicate with a particular project.

Consider this example
```sh
mkdir temp && cd temp

# Initialize the current directory with an Appwrite project 
appwrite init

temp 
└── appwrite.json
```

A `appwrite.json` file will be associated with one project only whose ID will be visible in the `project` attribute. 

```json
{
    "project" : "",
    "functions": [
        {
          "name" : "My awesome function 💪",
          "path" : "./FunctionOne",
          "command" : "python main.py",
          "runtime": "python3.9",

>           // This section is useful when we allow cloud functions to be tested locally

Reviewer 2: JSON can't have comments 😉

Give your project a name:
Project X

Creating Project X...

Project X Created 👍
```

> We need to first add support to utopia-php/CLI to allow for single select and multi select commands to enable this feature.

### 🟢 The `appwrite deploy` command

The `appwrite deploy` command is a convenient wrapper around three main commands
* Create Function
* Create Tag
* Update Tag

```sh
# Usage 1
appwrite deploy functions
Which functions would you like to deploy? 
✅️ FunctionOne
✅️ FunctionTwo
⭕️ FunctionThree

# Usage 2
appwrite deploy functions --all
# Deploys all the functions specified in appwrite.json

# Future support for static hosting
appwrite deploy site
appwrite deploy site --all 
```

## Easier creation of cloud functions
---

We will introduce an `appwrite generate` command that does the following
* Generate a function / static site template. 
* Make an entry in the `appwrite.json` file. Throw an error if the file doesn't exist.

Here are some example usages 
```sh
# Usage 1
appwrite generate function --name="FunctionOne" --runtime=python-3.9
# Creates a folder called FunctionOne with a template for a sample python function.
temp
├── appwrite.json
└── FunctionOne

>     ├── .appwrite

Reviewer 2: May be some information regarding `.appwrite` file will be useful as well

Reviewer 3: Is it possible to have everything in `appwrite.json`? I think it's a bit confusing to have two different types of config files.

Reviewer 4: Maybe most major configs are done using appwrite.json then .appwrite can be used for API keys and project keys as it'll be hidden by default right? Aslong as we remind users not to commit .appwrite they can share their project with ease

Reviewer 5: Agreeing with @kodumbeats - If we are going to have a `.appwrite` directory, why not have `config.yaml` in there (example, how .git is setup) or just have `appwrite.yaml`

Author: @lohanidamodar  `.appwrite` is actually the folder which some runtimes use to store their dependencies like Python and ruby  . Similar to node modules in npm projects

Author: @kodumbeats `appwrite.json` is a file that is expected to be a part of version control since it will have all the information related to a project like the collections, functions and so on and hence allow them to recreate their project environment . Having secrets in this file doesn't seem like a good idea

    ├── main.py
    └── requirements.txt

# Usage 2

> appwrite generate function --name="FunctionTwo" --runtime=node-16.0

Reviewer 2: May be we can introduce one more parameter here `--template` which should be a link to an accessible github repository/branch so that devs can customize to use their own templates as well as our more advance templates, even our demos can be used as template easily using this feature?

- Ways to scale the solution

Ensure that you include examples, code-snippets etc. to allow the community to understand the proposed solution. **It would be best if the examples use naming conventions that you intend to use during the actual implementation so that changes can be suggested early on during the development.**

Write your answer below.

-->

This refactor of the CLI will revolve around the following areas
1. Migration of the CLI from a Server Side SDK to a Console SDK
2. Easier deployment of Cloud Functions
3. Easier creation of Cloud Functions

## Migration of the CLI from a Server SDK to a Console SDK
---
 
Until now, the CLI was based off of our Server Side Swagger Spec. While this was a good start, it limits our functionality to a Server SDK. Ideally we want the CLI to be an alternative to the Console SDK ( aka the Appwrite Dashboard ). Here are some changes that we will need to make to adhere to the server spec.

After the implementation of this RFC, the CLI will behave more like the Appwrite Console rather than a Server SDK. This accounts for a new form of Authentication that is similar to the way we handle logins in the Appwrite console. We use a cookie.

> We will introduce a new command `appwrite login` that will work in one of either ways.

Reviewer 3: Do we need to pick one or the other? I understand that we should think to support OAuth login, but it's not a great CLI experience if a browser is required to login.

```sh
appwrite functions createTag --functionId=[ID] --command=[COMMAND]
```

This corresponding appwrite issue will also need to be tackled as we implement this.
https://github.com/appwrite/appwrite/issues/1316

### 🟢 The `appwrite.json` File

The next improvement is going to be along the lines of a `appwrite.json` file. Running `appwrite init` in the current directory will create a `appwrite.json` file and initialize the current directory with attributes necessary to communicate with a particular project.

Consider this example
```sh
mkdir temp && cd temp

# Initialize the current directory with an Appwrite project 
appwrite init

temp 
└── appwrite.json
```

A `appwrite.json` file will be associated with one project only whose ID will be visible in the `project` attribute. 

```json
{
    "project" : "",
    "functions": [
        {
          "name" : "My awesome function 💪",
          "path" : "./FunctionOne",
          "command" : "python main.py",
          "runtime": "python3.9",
          // This section is useful when we allow cloud functions to be tested locally
          "vars" :{
            "KEY 1" : "VALUE 1",
          }
        },
    ],
    "hosting": []

> }

Reviewer 5: I would like to suggest YAML or even TOML, support for them everywhere, easier to read and less prone to type-o frustration with quotes and such. Also... comments.

```

There are two scenarios here 
* A developer would like to link the directory to an existing project

```sh
appwrite init

> 

Reviewer 5: its this to add the Appwrite stack, to setup for functions or just create the config file with the project id? If project, will it also create the project for us on our endpoints?

How would you like to start? 
⭕️ Create a new project
✅️ Link to an existing project

Choose a project to link this directory to 
⭕️ Project 1 (nowfoms)
✅️ Project 2 (dfsdda)
⭕️ Project 3 (wedfdd)
```

* A developer would like to create a new project while initializing this directory

```sh
appwrite init

How would you like to start? 
✅️ Create a new project
⭕️ Link to an existing project

Give your project a name:
Project X

Creating Project X...

Project X Created 👍
```

Once you choose the project, appwrite.json is initialized with the newly created project's ID?

```json
{
  "projectId" : "xyz"
}
```

> We need to first add support to utopia-php/CLI to allow for single select and multi select commands to enable this feature.

### 🟢 The `appwrite deploy` command

The `appwrite deploy` command is a convenient wrapper around three main commands
* Create Function
* Create Tag
* Update Tag to activate it 

```sh
# Usage 1

> appwrite deploy functions --confirm 

Reviewer 6: For simplicity of use, one of `--confirm` or `--all` could be a default

Reviewer 3: This almost feels like an `-i --interactive`mode, like an interactive rebase in git?

Reviewer 5: could also follow the common CLI trends with `-y` instead of a --confirm or --all

Which functions would you like to deploy? 
✅️ FunctionOne
✅️ FunctionTwo
⭕️ FunctionThree

Deploying FunctionOne.. 
FunctionOne already exists 
Updating FunctionOne with new tag
Do you want to activate FunctionOne? [Yes/No]

# Usage 2
appwrite deploy functions --all
# Deploys all the functions specified in appwrite.json

> 

Reviewer 6: What do you think about adding a 4th usage for CI environments to deploy _some_ of the projects functions:

```bash
appwrite deploy functions --names "My awesome function 💪, FunctionTwo" or
appwrite deploy functions --names "My awesome function 💪" "FunctionTwo"
```

Reviewer 5: With this and others, I believe we should name space the commands `functions deploy` (and looking further down, `site deploy -y`

# Future support for static hosting
appwrite deploy site
appwrite deploy site --all 
```

## Easier creation of cloud functions
---

We will introduce an `appwrite init function` command that does the following
* Generate a function / static site template. 
* Make an entry in the `appwrite.json` file. Throw an error if the file doesn't exist.

Here are some example usages 
```sh
# Usage 1

> appwrite init function --functionId="abc" --name="FunctionOne" --runtime=python-3.9

Reviewer 6: Maybe we could allow `runtime` to be a language and use the latest version available by default? e.g.:

```bash
appwrite init function --functionId="abc" --name="FunctionOne" --runtime=python
```

Which would use `python-3.9` or whichever is newest at the time

Reviewer 5: Like above: `appwrite function init` -- This helps when printing help screens too, because you can. order them by the namespaces for readable grouping

# Creates a folder called FunctionOne with a template for a sample python function.
temp
├── appwrite.json
└── FunctionOne
    ├── .appwrite
    ├── main.py
    └── requirements.txt

# Usage 2
appwrite init function --functionId="xyz" --name="FunctionTwo" --runtime=node-16.0
# Creates a folder called FunctionTwo with a template for a sample node function.
temp 
├── appwrite.json
├── FunctionOne
│   ├── .appwrite
│   ├── main.py
│   └── requirements.txt
└── FunctionTwo
    ├── index.js
    ├── node_modules
    └── package.json

# Usage 3
appwrite init function

> Give your function an ID

Reviewer 3: Is it possible to validate this ID with the Key validator as a part of the CLI? We could provide real-time feedback even before a request is made.

class Config extends Reader {

  function getFunction(string $name): array {}

  function getFunctions(): array {}
  
  function setFunction(array $function): void {}

  function getProject(): string {}

  function setProject(string $id): void {}

}
```

### Prior art

[prior-art]: #prior-art

<!--

Discuss prior art, both the good and the bad, in relation to this proposal. A
few examples of what this can include are:

- Does this functionality exist in other software and what experience has their
  community had?
- For other teams: What lessons can we learn from what other communities have
  done here?
- Papers: Are there any published papers or great posts that discuss this? If
  you have some relevant papers to refer to, this can serve as a more detailed
  theoretical background.

This section is intended to encourage you as an author to think about the
lessons from other software, provide readers of your RFC with a fuller picture.
If there is no prior art, that is fine - your ideas are interesting to us
whether they are brand new or if it is an adaptation from other software.

Write your answer below.
-->

This RFC is inspired by the Firebase CLI.

> 

Reviewer 3: Does the Fission CLI have any insights to offer?
https://docs.fission.io/docs/usage/function/functions/
